# HG changeset patch
# User Adam Borowski <kilobyte@angband.pl>
# Date 1513514181 -3600
#      Sun Dec 17 13:36:21 2017 +0100
# Node ID a67340471339f947a278aedeff8af70c7e5248e9
# Parent  8854eb2c9fb4213835ad9c289181b4f0c8d64f67
manually touch atime when reading a mbox file

The only common use of atime left is local mail agents leaving a mark on
mbox file after the mail has been read.  And, since POSIX-2008, it is
possible to use futimens() to alter atime even on filesystems mounted with
noatime.

There's no extra cost for doing this on when atime updates are enabled:
the inode will be dirty already.

diff -r 8854eb2c9fb4 -r a67340471339 configure.ac
--- a/configure.ac	Sun Oct 15 15:13:18 2017 +0800
+++ b/configure.ac	Sun Dec 17 13:36:21 2017 +0100
@@ -447,6 +447,9 @@ AC_CHECK_FUNCS(ftruncate, , [AC_CHECK_LI
 dnl SCO has strftime() in libintl
 AC_CHECK_FUNCS(strftime, , [AC_CHECK_LIB(intl, strftime)])
 
+dnl Set the atime of files
+AC_CHECK_FUNCS(futimens)
+
 dnl AIX may not have fchdir()
 AC_CHECK_FUNCS(fchdir, , [mutt_cv_fchdir=no])
 
diff -r 8854eb2c9fb4 -r a67340471339 main.c
--- a/main.c	Sun Oct 15 15:13:18 2017 +0800
+++ b/main.c	Sun Dec 17 13:36:21 2017 +0100
@@ -392,6 +392,12 @@ static void show_version (void)
 #else
 	"-HAVE_RESIZETERM  "
 #endif
+	
+#ifdef HAVE_FUTIMENS
+	"+HAVE_FUTIMENS  "
+#else
+	"-HAVE_FUTIMENS  "
+#endif
         );	
   
   puts (
diff -r 8854eb2c9fb4 -r a67340471339 mbox.c
--- a/mbox.c	Sun Oct 15 15:13:18 2017 +0800
+++ b/mbox.c	Sun Dec 17 13:36:21 2017 +0100
@@ -440,6 +440,7 @@ static int mbox_open_mailbox (CONTEXT *c
     rc = mmdf_parse_mailbox (ctx);
   else
     rc = -1;
+  mutt_touch_atime (fileno (ctx->fp));
 
   mbox_unlock_mailbox (ctx);
   mutt_unblock_signals ();
@@ -1261,6 +1262,8 @@ int mutt_reopen_mailbox (CONTEXT *ctx, i
     return (-1);
   }
 
+  mutt_touch_atime (fileno (ctx->fp));
+
   /* now try to recover the old flags */
 
   index_hint_set = (index_hint == NULL);
diff -r 8854eb2c9fb4 -r a67340471339 muttlib.c
--- a/muttlib.c	Sun Oct 15 15:13:18 2017 +0800
+++ b/muttlib.c	Sun Dec 17 13:36:21 2017 +0100
@@ -1942,6 +1942,16 @@ void mutt_set_mtime (const char* from, c
   }
 }
 
+/* set atime to current time, just as read() would do on !noatime.
+ * Silently ignored if unsupported. */
+void mutt_touch_atime (int f)
+{
+#ifdef HAVE_FUTIMENS
+  struct timespec times[2]={{0,UTIME_NOW},{0,UTIME_OMIT}};
+  futimens(f, times);
+#endif
+}
+
 const char *mutt_make_version (void)
 {
   static char vstring[STRING];
diff -r 8854eb2c9fb4 -r a67340471339 protos.h
--- a/protos.h	Sun Oct 15 15:13:18 2017 +0800
+++ b/protos.h	Sun Dec 17 13:36:21 2017 +0100
@@ -122,6 +122,7 @@ time_t mutt_local_tz (time_t);
 time_t mutt_mktime (struct tm *, int);
 time_t mutt_parse_date (const char *, HEADER *);
 int is_from (const char *, char *, size_t, time_t *);
+void mutt_touch_atime (int);
 
 const char *mutt_attach_fmt (
 	char *dest,
